webservices/common/util.py

```diff
@@ -45,3 +47,47 @@ def get_class_by_tablename(tablename):
     for c in db.Model._decl_class_registry.values():
         if hasattr(c, '_sa_class_manager') and c._sa_class_manager.mapper.mapped_table == tablename:
             return c
+
+
+
+def year_defaults(**kwargs):
+    """ We have two ways of picking time and both are now optional.
+
+    It would be nice to deprecate cycle for itemized endpoints that
+    have receipt date in the future, but we can keep it going now for
+    a smoother customer experience and try infer the most consistent
+    defaults.
+
+    Here are the conditions:
+        - if no date supplied default to the current 2-year period
+        - if a cycle is selected, that will provide sufficient date filtering
+        - if there is a min and max date check to make sure there isn't more than a 6 year period
+        - if cycle and only one min date or max-date, it adds the restriction with in the 2-year period
+    """
+    min_date = kwargs.get('min_date')
+    max_date = kwargs.get('max_date')
+    cycle = kwargs.get('two_year_transaction_period')
+
+    if min_date is None and max_date is None and cycle is None:
+        kwargs['two_year_transaction_period'] = CURRENT_CYCLE
+        return kwargs
+    if cycle is not None:
+        return kwargs
+    # this is for cases where only one date (min_date or max_date) is present
+    if not min_date:
+        begin_cycle = max_date.year - max_date.year % 2
+        kwargs['min_date'] = datetime(begin_cycle, 1, 1)
+        return kwargs
+    if not max_date:
+        cycle = min_date.year + min_date.year % 2
+        kwargs['max_date'] = datetime(cycle, 12, 31)
+        return kwargs
+    # We plan on rolling this back the 6 year restriction in the future
+    # but wanted to confirm performance on 6 years first.
+    if min_date and max_date:
+        if (max_date - min_date).days > 2190:
```

Comment:
I was going to ask if it'd be possible do `(max_date - min_date).years` instead for clarity, but the resulting `datetime.timedelta` object doesn't have that property, d'oh!  Python, you've let me down! 😆

Reply:
After chatting with some others about this, we might be able to leverage the [`dateutil.relativedelta`](https://dateutil.readthedocs.io/en/stable/relativedelta.html) object (from the `dateutil` library) to do this, as even going with the exact days difference ought to probably account for things like leap days and whatnot...  GitHub, you need a :shakes_fist: emoji. :-)

Reply:
Let's just add 2 since that is as many leap years you can have in a 6 year period. As you know, we don't really have to be scientifically exact on this, it's just our guard rails while we try out our indexes.